src/expr/src/vector_op/agg/array_agg.rs

```diff
@@ -0,0 +1,262 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::array::{ArrayBuilder, ArrayBuilderImpl, DataChunk, ListValue, RowRef};
+use risingwave_common::error::{ErrorCode, Result};
+use risingwave_common::types::{DataType, Datum, Scalar};
+use risingwave_common::util::ordered::OrderedRow;
+use risingwave_common::util::sort_util::{OrderPair, OrderType};
+
+use crate::vector_op::agg::aggregator::Aggregator;
+
+#[derive(Clone)]
+struct ArrayAggUnordered {
+    return_type: DataType,
+    agg_col_idx: usize,
+    values: Vec<Datum>,
+}
+
+impl ArrayAggUnordered {
+    fn new(return_type: DataType, agg_col_idx: usize) -> Self {
+        debug_assert!(matches!(return_type, DataType::List { datatype: _ }));
+        ArrayAggUnordered {
+            return_type,
+            agg_col_idx,
+            values: vec![],
+        }
+    }
+
+    fn push(&mut self, datum: Datum) {
+        self.values.push(datum);
+    }
+
+    fn get_result_and_reset(&mut self) -> ListValue {
+        ListValue::new(std::mem::take(&mut self.values))
+    }
+}
+
+impl Aggregator for ArrayAggUnordered {
+    fn return_type(&self) -> DataType {
+        self.return_type.clone()
+    }
+
+    fn update_single(&mut self, input: &DataChunk, row_id: usize) -> Result<()> {
+        let array = input.column_at(self.agg_col_idx).array_ref();
+        self.push(array.datum_at(row_id));
+        Ok(())
+    }
+
+    fn update_multi(
+        &mut self,
+        input: &DataChunk,
+        start_row_id: usize,
+        end_row_id: usize,
+    ) -> Result<()> {
+        for row_id in start_row_id..end_row_id {
```

Comment:
How about reserving some capacity for `self.values` before pushing one-by-one?

Reply:
ok